pkg/rcmanager/ipinstance.go

```diff
@@ -0,0 +1,210 @@
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/gogf/gf/container/gset"
+	networkingv1 "github.com/oecp/rama/pkg/apis/networking/v1"
+	"github.com/oecp/rama/pkg/constants"
+	"github.com/oecp/rama/pkg/utils"
+	v1 "k8s.io/api/core/v1"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog"
+)
+
+// reconcile one single node, update/add/remove everything about it's
+// corresponding remote vtep.
+func (m *Manager) reconcileIPInstance(nodeName string) error {
+	klog.Infof("[remote cluster] Starting reconcile ipinstance from cluster %v, node name=%v", m.ClusterName, nodeName)
+	if len(nodeName) == 0 {
+		return nil
+	}
+	var (
+		node     *v1.Node
+		err      error
+		vtepName string
+	)
+	vtepName = utils.GenRemoteVtepName(m.ClusterName, nodeName)
+	node, err = m.NodeLister.Get(nodeName)
+	if err != nil {
+		if k8serror.IsNotFound(err) {
+			err = m.LocalClusterRamaClient.NetworkingV1().RemoteVteps().Delete(context.TODO(), vtepName, metav1.DeleteOptions{})
+			if k8serror.IsNotFound(err) {
+				return nil
+			}
+		}
+		return err
+	}
+
+	var (
+		vtepIP     = node.Annotations[constants.AnnotationNodeVtepIP]
+		vtepMac    = node.Annotations[constants.AnnotationNodeVtepMac]
+		newVtep    = false
+		remoteVtep *networkingv1.RemoteVtep
+	)
+
+	remoteVtep, err = m.RemoteVtepLister.Get(vtepName)
+	if err != nil {
+		if !k8serror.IsNotFound(err) {
+			return err
+		}
+		remoteVtep = utils.NewRemoteVtep(m.ClusterName, m.RemoteClusterUID, vtepIP, vtepMac, node.Name, nil)
+		newVtep = true
+	}
+
+	var (
+		curTime    = metav1.Now()
+		desired    []string
+		desiredSet *gset.StrSet
+		actual     []string
+		actualSet  *gset.StrSet
+		remove     *gset.StrSet
+		add        *gset.StrSet
+		// vtepIP or vtepMac changed
+		vtepChanged           bool
+		endpointILListChanged bool
+	)
+
+	desired, err = m.pickEndpointListFromNode(node)
+	if err != nil {
+		return err
+	}
+	actual = remoteVtep.Spec.EndpointIPList
+	desiredSet = gset.NewStrSetFrom(desired)
+	actualSet = gset.NewStrSetFrom(actual)
+	remove = actualSet.Diff(desiredSet)
+	add = desiredSet.Diff(actualSet)
+
+	remoteVtep = remoteVtep.DeepCopy()
+	remoteVtep.Spec.VtepIP = vtepIP
+	remoteVtep.Spec.VtepMAC = vtepMac
+	remoteVtep.Spec.EndpointIPList = desiredSet.Slice()
+	remoteVtep.Status.LastModifyTime = curTime
+
+	vtepChanged = vtepIP != "" && vtepMac != "" && (vtepIP != remoteVtep.Spec.VtepIP || vtepMac != remoteVtep.Spec.VtepMAC)
+	endpointILListChanged = remove.Size() != 0 || add.Size() != 0
+	if !newVtep && !endpointILListChanged && !vtepChanged {
+		return nil
+	}
+	if newVtep {
+		remoteVtep, err = m.LocalClusterRamaClient.NetworkingV1().RemoteVteps().Create(context.TODO(), remoteVtep, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+	} else {
+		remoteVtep, err = m.LocalClusterRamaClient.NetworkingV1().RemoteVteps().Update(context.TODO(), remoteVtep, metav1.UpdateOptions{})
```

Comment:
It seems remote vtep CRs will be updated both in node reconciling and ipinstance reconciling, will there be any conflict if the same remote vtep CR is updated at the same time? (There might be, because both of the two places are updating a whole remote vtep CR)

Reply:
ditto.